azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -190,8 +191,50 @@ public Props getProps() {
   }
 
   public String getEffectiveUser() {
-    return this.props.getString(JobProperties.USER_TO_PROXY,
-        this.getNode().getExecutableFlow().getSubmitUser());
+    return this.effectiveUser;
+  }
+
+  /**
+   * Determine the effective user for the job.
+   * <p>
+   * If the jobType for this job has a defaultProxyUser then return that. If the user.to.proxy is
+   * not specified in the job props, then return submit user. If the user.to.proxy is specified in
+   * the Job, and - It is also part of list of proxy users in the project permissions, then return
+   * the same. - Otherwise return Optional.empty()
+   *
+   * @return {@link Optional} effective user for the Job
+   */
+  public Optional<String> determineEffectiveUser() {
+    Optional<String> jobType = JobTypeManager.getJobType(this.props);
+    if (jobType.isPresent()) {
+      // JobTypePluginSet is never NULL
+      Optional<String> defaultProxyUser = this.jobtypeManager.getJobTypePluginSet()
+          .getDefaultProxyUser(jobType.get());
+      if (defaultProxyUser.isPresent()) {
+        this.logger.info(String.format("Found default proxy user %s for the job type %s",
+            defaultProxyUser.get(), jobType.get()));
+        return defaultProxyUser;
+      }
+    }
```

Comment:
Should this be factored out into its own method?

Reply:
There is not much calculation to it apart from returning the value and logging it, hence refactoring may not be needed here. The actual calculation of defaultProxyUser is already part of the method azkaban.jobtype.JobTypePluginSet#getDefaultProxyUser

Reply:
Fair enough.